SCClassLibrary/Common/Collections/Dictionary.sc

```diff
@@ -560,12 +560,32 @@ IdentityDictionary : Dictionary {
 		^this.superPerformList(\doesNotUnderstand, selector, args);
 	}
 
-		// Quant support.
-		// The Quant class assumes the quant/phase/offset scheduling model.
-		// If you want a different model, you can write a dictionary like so:
-		// (nextTimeOnGrid: { |self, clock| ... calculate absolute beat number here ... },
-		//	parameter: value, parameter: value, etc.)
-		// If you leave out the nextTimeOnGrid function, fallback to quant/phase/offset.
+	doesNotUnderstandWithKeys {|selector, argsArray, keywordArgsAsPairs|
+		if(know.not){
+			^this.superPerformList(\doesNotUnderstandWithKeys, selector, argsArray, keywordArgsAsPairs)
+		};
+
+		this[selector] !? {|f|
+			^f.functionPerformWithArgsAndKwArgs(\value, [this] ++ argsArray, keywordArgsAsPairs.asEvent)
+		};
+
+		// You cannot call a setter with a keyword arg, it does not make sense.
+
+		// If the keyword arg names don't match this will fail.
+		// Note how the selector is passed in here.
+		this[\forward] !? {|f|
+			^f.functionPerformWithArgsAndKwArgs(\value, [this, selector] ++ argsArray, keywordArgsAsPairs.asEvent)
```

Comment:
one more suggestion:

let's add a last `args` argument to `performWithEnvir` in both Object and Function, then the method `functionPerformWithArgsAndKwArgs` should nor be necessary.

What do you think?

Reply:
If I've understood you correctly — as far as I can see, classes have to opt into being called through object prototyping by implementing  `functionPerformList`, otherwise they just return themselves, meaning replacing this with `performWithEnvir` would be a breaking change. 

It would not be possible to construct an argument array from args and kwargs from inside identity dictionary as you'd have to know what the held object did inside `functionPerformList`,
and `functionPerformList` uses an array of args, meaning you can't get the argnames either.

Without breaking object prototyping, I can't see another way (unless I've misunderstood you).

If the interface for prototyping was to change though, one option would be for `doesNotUnderstand*` to check the first argument (second argument if it is 'this') of the object's `value` method (or FunctionDef if function) and if it is called 'self' then pass the dictionary in. This way, `functionPerformList` could be remove entirely, and any class or function can opt into this behaviour. While switch on argument names is perhaps odd, I don't think there is any other use for the word 'self' in supercollider.